com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/ResolveClasspathsHandler.java

```diff
@@ -83,36 +88,104 @@ private static IJavaProject getJavaProjectFromName(String projectName) throws Co
      *             CoreException
      */
     public static List<IJavaProject> getJavaProjectFromType(String fullyQualifiedTypeName) throws CoreException {
+        // If only one Java project exists in the whole workspace, return the project directly.
+        List<IJavaProject> javaProjects = JdtUtils.listJavaProjects(ResourcesPlugin.getWorkspace().getRoot());
+        if (javaProjects.size() <= 1) {
+            return javaProjects;
+        }
+
         String[] splitItems = fullyQualifiedTypeName.split("/");
         // If the main class name contains the module name, should trim the module info.
         if (splitItems.length == 2) {
             fullyQualifiedTypeName = splitItems[1];
         }
         final String moduleName = splitItems.length == 2 ? splitItems[0] : null;
 
+        // Search the associated project for the type.
+        List<IJavaProject> projects = searchJavaProjectFromType(fullyQualifiedTypeName, moduleName);
+        if (projects.isEmpty()) {
+            String className = fullyQualifiedTypeName.substring(fullyQualifiedTypeName.lastIndexOf('.') + 1);
+            // See the bug https://github.com/Microsoft/vscode-java-debug/issues/427
+            // If the target class is an inner class, its fully qualified name will contain $, but the search engine doesn't support $ separated class name.
+            // Add an extra search from main method.
+            if (className.indexOf('$') > 0) {
+                return searchJavaProjectFromMainMethod(fullyQualifiedTypeName, moduleName);
```

Comment:
searchJavaProjectFromMainMethod in SearchEngine.createWorkspaceScope is not a good idea.